### PR TITLE
Support knobs

### LINKED ISSUE #14: Support handle/knob
I think the title does it :-)

### PR DESCRIPTION
**Pull request by phaistonian**
This PR Closes #14 

### Notes:

- I am not into TypeScript, so you might want to clean up the code a bit.
- This target checking against the knobs should also be applied to the touch event - missed that.
- README could also need an update.

https://user-images.githubusercontent.com/796530/124083708-dfb29f00-da56-11eb-9d42-bc3c74b1ddd5.mp4

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -29,6 +31,8 @@ const SortableList = ({ children, allowDrag = true, onSortEnd, draggedItemClassN
   const itemsRef = React.useRef<HTMLElement[]>([])
   // this array contains the coordinates of each sortable element (only computed on dragStart and used in dragMove for perf reason)
   const itemsRect = React.useRef<DOMRect[]>([])
+  // Hold all registered knobs
+  const knobs = React.useRef([]);
```
Comment (ValentinH) [inline src/index.tsx:35]:
```suggestion
  const knobs = React.useRef<HTMLElement[]>([]);
```

File: src/hooks.ts [hunk 2]
```diff
@@ -144,6 +145,11 @@ export const useDrag = ({ onStart, onMove, onEnd, containerRef }: UseDragProps)
         // we don't want to handle clicks other than left ones
         return
       }
+
+      if (knobs.length && !knobs.find(knob => knob.contains(e.target))) {
```
Comment (ValentinH) [inline src/hooks.ts:149]: As you mentioned, this should be done for the onTouchStart event as well.

File: src/hooks.ts [hunk 3]
```diff
@@ -39,9 +39,10 @@ type UseDragProps = {
   onMove?: (args: OnMoveArgs) => void
   onEnd?: () => void
   containerRef: React.MutableRefObject<HTMLDivElement | null>
+  knobs?: [],
```
Comment (ValentinH) [inline src/hooks.ts:42]:
```suggestion
  knobs?: HTMLElement[],
```

File: src/index.tsx [hunk 4]
```diff
@@ -273,3 +290,32 @@ export const SortableItem = ({ children }: ItemProps) => {
 
   return React.cloneElement(children, { ref: elementRef })
 }
+
+export const SortableKnob = ({ children  }: ItemProps) => {
+  const context = React.useContext(SortableListContext)
+
+  if (!context) {
+    throw new Error('SortableItem must be a child of SortableList')
```
Comment (ValentinH) [inline src/index.tsx:298]:
```suggestion
    throw new Error('SortableKnob must be a child of SortableList')
```

File: src/stories/simple-vertical-list/index.stories.tsx [hunk 5]
```diff
@@ -1,15 +1,17 @@
+
```
Comment (ValentinH) [inline src/stories/simple-vertical-list/index.stories.tsx:1]: I think we should create a dedicated story for knobs 🙂

**Review by ValentinH — CHANGES_REQUESTED** [review]
Thanks for the nice contribution! 🎉 

Please see my review comments, hopefully, it should be quite short to adapt

File: src/hooks.ts (hunk 2 above)
Comment (phaistonian) [reply src/hooks.ts:149]: Done :)

File: src/stories/simple-vertical-list/index.stories.tsx (hunk 5 above)
Comment (phaistonian) [reply src/stories/simple-vertical-list/index.stories.tsx:1]: Your call :)

**Comment by phaistonian** [comment]
As a side-note, for this to be more efficient in terms of performance, knobs should be promoted to a map and have knobs per item, i.e `registerKnob` = `knobsMap.get(itemRef).push(knob)`.

It's not a biggie, but in super-huge lists, it might have some impact.

File: README.md [hunk 6]
```diff
@@ -59,7 +59,10 @@ const App = () => {
     <SortableList onSortEnd={onSortEnd} className="list" draggedItemClassName="dragged">
       {items.map((item) => (
         <SortableItem key={item}>
-          <div className="item">{item}</div>
+          <div className="item">
```
Comment (ValentinH) [inline README.md:62]: I actually would not show the `SortableKnob` as part of the `basic example`. Let me add another example in the `SortableKnow` section.

**Review by ValentinH — APPROVED** [review]

**Comment by ValentinH** [comment]
Thanks for this great contribution! This was released with https://github.com/ricardo-ch/react-easy-sort/releases/tag/v1.2.0.

**Pull request merged**